rasa/cdu/command_processor.py

```diff
@@ -0,0 +1,270 @@
+
+
+def _slot_sets_after_latest_message(tracker: DialogueStateTracker) -> List[SlotSet]:
+    """Get all slot sets after the latest message.
+
+    Args:
+        tracker: The tracker to get the slot sets from.
+
+    Returns:
+    All slot sets after the latest user message.
+    """
+    if not tracker.latest_message:
+        return []
+
+    slot_sets = []
+
+    for event in reversed(tracker.applied_events()):
+        if isinstance(event, UserUttered):
+            break
+        elif isinstance(event, SlotSet):
+            slot_sets.append(event)
+    return slot_sets
+
+
+def _get_commands_from_tracker(tracker: DialogueStateTracker) -> List[Command]:
+    """Extracts the commands from the tracker.
+
+    Args:
+        tracker: The tracker containing the conversation history up to now.
+
+    Returns:
+    The commands.
+    """
+    if tracker.latest_message:
+        dumped_commands = tracker.latest_message.parse_data.get(COMMANDS) or []
+        assert isinstance(dumped_commands, list)
+        return [command_from_json(command) for command in dumped_commands]
+    else:
+        return []
+
+
+def execute_commands(
+    tracker: DialogueStateTracker, all_flows: FlowsList
+) -> List[Event]:
+    """Executes a list of commands.
+
+    Args:
+        commands: The commands to execute.
+        tracker: The tracker to execute the commands on.
+        all_flows: All flows.
+
+    Returns:
+    A tuple of the action to execute and the events that were created.
+    """
+    commands: List[Command] = _get_commands_from_tracker(tracker)
+    flow_stack = FlowStack.from_tracker(tracker)
+    original_stack_dump = flow_stack.as_dict()
+
+    current_top_flow = flow_stack.top_flow(all_flows)
+
+    commands = clean_up_commands(commands, tracker, all_flows)
+
+    events: List[Event] = []
+    collected_corrections = []
+    # TODO: should this really be reversed? 🤔
+    reversed_commands = list(reversed(commands))
+    for i, command in enumerate(reversed_commands):
+        if isinstance(command, CorrectSlotCommand):
+            structlogger.debug("command_executor.correct_slot", command=command)
+            collected_corrections.append(command)
+            # pulling in all subsequent correction commands into a single correction
+            if i < (len(reversed_commands) - 1) and \
+                isinstance(reversed_commands[i+1], CorrectSlotCommand):
+                continue
+            for correction in collected_corrections:
+                events.append(SlotSet(correction.name, correction.value))
+            events.append(SlotSet(CORRECTED_SLOTS_SLOT,
+                                  [s.name for s in collected_corrections]))
+            flow_stack.push(
+                FlowStackFrame(
+                    flow_id=FLOW_PATTERN_CORRECTION_ID,
+                    frame_type=StackFrameType.CORRECTION,
+                )
+            )
+            collected_corrections = []
```

Comment:
There's more issues here with corrections
My current changes address to following issues:
* Jump back works again, by setting the `CORRECTED_SLOTS_SLOT` again to the corrected slots. Extracting from the tracker did not work anymore, because slots have not been set at this point
* Preventing double corrections by fusing them into a single correction. This can happen in the restaurant flow when e.g. time and date are both adjusted. Also fixes a bug in this limited scenario, which still exists otherwise, see below.

HOWEVER
There's still more that's problematic here:
* When we get multiple disconnected correction groups, we are probably going to overwrite the `CORRECTED_SLOTS_SLOT` and only one correction group survives
* We probably need to order the slot sets somewhere so that we don't get disconnections by accident (e.g. Correct(A), SetSlot(B), Correct(C) -> Correct(A), Correct(C), SetSlot(B))
* I don't really know what to do if you would get corrections for slots of different flows and such scenarios. This used to be TOO_COMPLEX